integrations/amazon_bedrock/src/haystack_integrations/components/generators/amazon_bedrock/chat/adapters.py

```diff
@@ -0,0 +1,258 @@
+
+        # Truncate prompt if prompt tokens > model_max_length-max_length
+        # (max_length is the length of the generated text)
+        # TODO use Anthropic tokenizer to get the precise prompt length
+        # See https://github.com/anthropics/anthropic-sdk-python?tab=readme-ov-file#token-counting
+        self.prompt_handler = DefaultPromptHandler(
+            tokenizer="gpt2",
+            model_max_length=model_max_length,
+            max_length=self.generation_kwargs.get("max_tokens_to_sample") or 512,
+        )
+
+    def prepare_body(self, messages: List[ChatMessage], **inference_kwargs) -> Dict[str, Any]:
+        default_params = {
+            "max_tokens_to_sample": self.generation_kwargs.get("max_tokens_to_sample") or 512,
+            "stop_sequences": ["\n\nHuman:"],
+        }
+
+        # combine stop words with default stop sequences, remove stop_words as Anthropic does not support it
+        stop_sequences = inference_kwargs.get("stop_sequences", []) + inference_kwargs.pop("stop_words", [])
+        if stop_sequences:
+            inference_kwargs["stop_sequences"] = stop_sequences
+        params = self._get_params(inference_kwargs, default_params)
+        body = {"prompt": self.prepare_chat_messages(messages=messages), **params}
+        return body
+
+    def prepare_chat_messages(self, messages: List[ChatMessage]) -> str:
+        conversation = []
+        for index, message in enumerate(messages):
+            if message.is_from(ChatRole.USER):
+                conversation.append(f"{AnthropicClaudeChatAdapter.ANTHROPIC_USER_TOKEN} {message.content.strip()}")
+            elif message.is_from(ChatRole.ASSISTANT):
+                conversation.append(f"{AnthropicClaudeChatAdapter.ANTHROPIC_ASSISTANT_TOKEN} {message.content.strip()}")
+            elif message.is_from(ChatRole.FUNCTION):
+                error_message = "Anthropic does not support function calls."
+                raise ValueError(error_message)
+            elif message.is_from(ChatRole.SYSTEM) and index == 0:
+                # Until we transition to the new chat message format system messages will be ignored
+                # see https://docs.anthropic.com/claude/reference/messages_post for more details
+                logger.warning(
+                    "System messages are not fully supported by the current version of Claude and will be ignored."
+                )
+            else:
+                invalid_role = f"Invalid role {message.role} for message {message.content}"
+                raise ValueError(invalid_role)
+
+        prepared_prompt = "".join(conversation) + AnthropicClaudeChatAdapter.ANTHROPIC_ASSISTANT_TOKEN + " "
+        return self._ensure_token_limit(prepared_prompt)
+
+    def check_prompt(self, prompt: str) -> Dict[str, Any]:
+        return self.prompt_handler(prompt)
+
+    def _extract_messages_from_response(self, response_body: Dict[str, Any]) -> List[ChatMessage]:
```

Comment:
Would it make sense to have this function define the parent class with an extra param saying which entry in the dictionary contains the message, i.e., it's not metadata?

Example:

```
    def _extract_messages_from_response(self, message_tag: str, response_body: Dict[str, Any]) -> List[ChatMessage]:
        metadata = {k: v for (k, v) in response_body.items() if k != message_tag}
        return [ChatMessage.from_assistant(response_body[message_tag], meta=metadata)]

```

This is just to avoid having to define dysfunction for every `BedrockModelChatAdapter`

Reply:
smart 👍